### ISSUE TITLE
Feature Request: Exception Handler Keyboard Shortcuts

### ISSUE THREAD

**Issue by blinklad** [MolecularMatters/lpp_public#7]
# Feature Request: Exception Handler Keyboard Shortcuts

Firstly and if this is documented on the website, or feature requests of this type are universally binned as "won't fix", then disregard this issue and mark it as closed.

## Motivation

My codebase is judiciously littered with assertions. In builds with hot-loading enabled, the most top-level handler is Live++ unless a debugger is attached.
If an assertion fires, I perform logging + UI feedback before crashing the application. At this point, Live++ intercepts the continuable exception and presents the dialog below. So far so good.

A very common workflow for me is to continuously iterate on a top-level, conditionally invoked procedure - think of a in-game command that triggers test code, presents a debug UI, or any other conditional sanity checks. If an assert fires here, I want to _very quickly_ return to the callee. Selecting the 'Leave Function` button can be a large factor of the time in tight iterations; even on my laptop, if I'm re-compiling from scratch, it might take a second to rebuild, link and patch the exe, whereas moving the cursor to the button is slow and clumsy.

In general, I would like to say Live++ really doesn't get in my way when it comes to rapid iteration times except in this instance.

### Option 0
In the picture below, offering a shortcut for 'Leave Function' at the very least would be highly useful for me. The others aren't so relevant for me here, just 'Leave Function'.

<img width="631" alt="Exception Handler pop-up window, with buttons highlighted red" src="https://github.com/MolecularMatters/lpp_public/assets/41582020/602b7230-4c2d-476b-a8dc-9f0b0fa3a648">

However, it's worth noting that if I configure Live++ to always pop-up when intercepting an exception, none of the controls are 'focused', so I can't just fire off the last selected control or what have you.

### Option 1
If it were possible to remember the last selected option in the current Broker session whenever the dialog pops up, that may be a clean, orthogonal way to implement what I'm asking, while proving a more accessible interface to boot.

**Comment by MolecularMatters** [comment]
Hi,

Thank you for the detailed request, this is absolutely the right place to ask for something like this.
I very much like your suggestion in Option 1, and I think that maybe both options could be done?

Will follow up on this once I had a better look at it.

**Comment by blinklad** [comment]
Excellent. I think Option 1 is the cleanest way.
However, I have a related issue if you're keen on going down the road of allowing focus on those pop-up controls.
There is a setting to focus the Broker window on one of `Error`, `Success`, `Hot-Reload `, `Never `and `Always`.
`Error ` here is any one of `Compile`, `Link `or` Exception Handler`.

The problem is that I have my own diagnostics that intercept Compile + Link errors and display those in-game using the Live++ Hook API, but I can't prevent the Broker from focusing on those errors still.

In the original scenario, I want to immediately continue on Exception via Leave Function, and use my own in-game diagnostics to deal with Compile + Link errors. Ergo, in my workflow, the Broker is only a useful pop-up when it comes to the Exception behavior. With the default broker behavior, I have to Alt+Tab to bring up the broker if I'm not using multiple monitors, and that doesn't always handle the case when the Broker is minimized to the system tray and will re-focus to the primary monitor.

Hope the above makes sense, thanks for the fast reply.

**Comment by MolecularMatters** [comment]
I started looking into this and it feels like we first have to tackle the behaviour of the Live++ exception dialog.

> However, it's worth noting that if I configure Live++ to always pop-up when intercepting an exception, none of the controls are 'focused', so I can't just fire off the last selected control or what have you.

Can you share your settings with me?
The exception hander dialog is supposed to open and then be focused, no matter the current notification settings.

It should also open in case the Broker is minimized in the system tray, so changing your notification settings to e.g. "Never" might be a general solution that lets you intercept Compile + Link errors using your hooks, while Live++ displays the dialog in case of an exception.

I'll look into assigning shortcuts to the controls and focusing the last selection.

**Comment by MolecularMatters** [comment]
![image](https://github.com/MolecularMatters/lpp_public/assets/61518974/be43568e-c424-41c8-9d5a-af4c3a6e2c1a)

I added shortcuts to each of the action buttons, and the dialog will now focus on the last selected action, or "Leave function" by default. I also renamed "Ignore exception" to "Continue execution" since it better explains what's happening and also allows us to assign non-ambiguous shortcuts.

If this works for you, I'll ship this with the next update.
